### PR TITLE
[JENKINS-47768] - Avoid having "authenticated" twice in the group membership of a user (LastGrantedAuthorities)

### PR DESCRIPTION
**Pull request by Wadeck**
- this occur when the SecurityRealm potentially already grants that role (like in github-oauth-plugin)

See [JENKINS-47768](https://issues.jenkins-ci.org/browse/JENKINS-47768).

### Submitter checklist

- [x] JIRA issue is well described
- [x] Appropriate autotests or explanation to why this change has no tests

### Desired reviewers

@reviewbybees

### CODE REVIEW

**Comment by Wadeck** [comment]
Instead of checking that on each "update" I think it's better to do on the information retrieval. The main advantage is that we do not bother performance for user that do not use the "cache" that is provided by LastGrantedAuthorities. I think it's the majority of cases.

**Review by oleg-nenashev — COMMENTED** [review]
Would it be possible to use Set or Map instead? Just to avoid N^2 complexity on this method. 

Maybe it's just better to modify the internal data structure and to use Maps in new API

File: core/src/main/java/jenkins/security/LastGrantedAuthoritiesProperty.java [hunk 1]
```diff
@@ -48,14 +48,20 @@ public UserProperty reconfigure(StaplerRequest req, JSONObject form) throws Form
...
-        r[0] = SecurityRealm.AUTHENTICATED_AUTHORITY;
-        if (roles != null) {
-            for (int i = 1; i < r.length; i++) {
-                r[i] = new GrantedAuthorityImpl(roles[i - 1]);
+        if(roles == null){
+            return new GrantedAuthority[]{SecurityRealm.AUTHENTICATED_AUTHORITY};
+        }
+
+        String authenticatedRole = SecurityRealm.AUTHENTICATED_AUTHORITY.getAuthority();
+        List<GrantedAuthority> grantedAuthorities = new ArrayList<>();
```
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/security/LastGrantedAuthoritiesProperty.java]: Initialize with `roles.length`

**Review by oleg-nenashev — APPROVED** [review]

File: core/src/main/java/jenkins/security/LastGrantedAuthoritiesProperty.java (hunk 1 above)
Comment (Wadeck) [reply core/src/main/java/jenkins/security/LastGrantedAuthoritiesProperty.java]: changed

File: core/src/main/java/jenkins/security/LastGrantedAuthoritiesProperty.java [hunk 2]
```diff
@@ -48,14 +48,20 @@ public UserProperty reconfigure(StaplerRequest req, JSONObject form) throws Form
...
+        if(roles == null){
+            return new GrantedAuthority[]{SecurityRealm.AUTHENTICATED_AUTHORITY};
+        }
+
+        String authenticatedRole = SecurityRealm.AUTHENTICATED_AUTHORITY.getAuthority();
+        List<GrantedAuthority> grantedAuthorities = new ArrayList<>(roles.length);
+        for (int i = 0; i < roles.length; i++){
+            // to avoid having twice that role
+            if(!authenticatedRole.equals(roles[i])){
+                grantedAuthorities.add(new GrantedAuthorityImpl(roles[i]));
```
Comment (dwnusbaum) [inline core/src/main/java/jenkins/security/LastGrantedAuthoritiesProperty.java]: If `this.roles` is not null, but doesn't contain `SecurityRealm.AUTHENTICATED_AUTHORITY`, then `SecurityRealm.AUTHENTICATED_AUTHORITY` won't be added to the result. Unless I'm missing something I think you want to add `AUTHENTICATED_AUTHORITY` to the the list before the loop to make sure it's included, right? (And initialize the list to `roles.length + 1`).

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
Agreed with @dwnusbaum, the test failures are related

**Review by dwnusbaum — APPROVED** [review]
🐝

**Review by oleg-nenashev — APPROVED** [review]
🐝 and 👍 for reworking just another Groovy test

**Comment by oleg-nenashev** [comment]
@reviewbybees done

**Pull request merged**